### PR TITLE
Sarah/correct normal

### PR DESCRIPTION
**Pull request by SarahRo**
This pull request addresses the following points
- use -n_y in contact formulation instead of normalised gap function
- fix gradual loading
- clean up python demos

### CODE REVIEW

File: cpp/Contact.hpp [hunk 1]
```diff
@@ -1350,6 +1316,119 @@ class Contact
     return {std::move(c), cstride};
   }
 
+  /// Compute inward surface normal at Pi(x)
+  /// @param[in] orgin_meshtag - surface on which to integrate
+  /// @param[out] c - normals ny packed on facets.
```
Comment (jorgensd) [inline cpp/Contact.hpp:1321]: Add specification of @param[gap]

File: cpp/Contact.hpp [hunk 2]
```diff
@@ -1350,6 +1316,119 @@ class Contact
...
+    xt::xtensor<double, 4> phi(cmap.tabulate_shape(1, 1));
+    for (int i = 0; i < num_facets; i++)
+    {
+      auto cell = (*puppet_facets)(i, 0); // extract cell
+
+      for (int j = 0; j < num_q_points; j++)
+      {
+        sorted_cells[j] = (*map)(i, j, 0);
+        // Is there a better way to do this???
+        for (int k = 0; k < gdim; k++)
```
Comment (jorgensd) [inline cpp/Contact.hpp:1385]: Is "this" Computing the the quadrature points on the other facet (i.e. summing quadrature point for a given facet with the gap)?

I think this function would be cleaner if it was not part of the contact class, and simply took in the cells and corresponding points on each cell, and outputted the normals.

Comment (jorgensd) [reply cpp/Contact.hpp:1385]: This "standalone" function could start around line 1360 i think

File: cpp/contact_kernels.hpp [hunk 3]
```diff
@@ -237,43 +239,23 @@ kernel_fn generate_contact_kernel(
...
       // if normal not constant, get surface normal at current quadrature point
+      int normal_offset = c_offset + offsets[5];
       if (!constant_normal)
       {
         n_dot = 0;
         for (int i = 0; i < gdim; i++)
         {
-          gap += c[gap_offset + q * gdim + i] * c[gap_offset + q * gdim + i];
-          n_surf(i) = c[gap_offset + q * gdim + i];
+          n_surf(i) = -c[normal_offset + q * gdim + i];
```
Comment (jorgensd) [inline cpp/contact_kernels.hpp:250]: Would be good to document why we use a minus here (and probably every other place we compute n_surf)

File: cpp/contact_kernels.hpp [hunk 4]
```diff
@@ -320,15 +302,15 @@ kernel_fn generate_contact_kernel(
         R_minus = 0;
       else
         R_minus = R_minus * detJ * weights[q];
-      sign_u *= -detJ * weights[q];
+      sign_u *= detJ * weights[q];
```
Comment (jorgensd) [inline cpp/contact_kernels.hpp:305]: What is the reason for sign change here and below?

File: cpp/contact_kernels.hpp [hunk 5]
```diff
@@ -393,24 +375,22 @@ kernel_fn generate_contact_kernel(
     xt::xarray<double> n_surf = xt::zeros<double>({gdim});
     // FIXME: Code duplication from previous kernel, and should be made into a
     // lambda function
+    double n_dot = 0;
     if (constant_normal)
     {
+      // If surface normal constant precompute (n_phys * n_surf)
       for (int i = 0; i < gdim; i++)
+      {
         n_surf(i) = w[i + 2];
```
Comment (jorgensd) [inline cpp/contact_kernels.hpp:384]: Here `n_surf` is not negative, why the difference from the previous kernel?

File: cpp/utils.hpp [hunk 6]
```diff
@@ -601,16 +601,13 @@ pack_circumradius_facet(std::shared_ptr<const dolfinx::mesh::Mesh> mesh,
...
 //-----------------------------------------------------------------------------
-xt::xtensor<double, 3>
-get_basis_functions(xt::xtensor<double, 3>& J, xt::xtensor<double, 3>& K,
-                    xt::xtensor<double, 1>& detJ,
-                    const xt::xtensor<double, 2>& x,
-                    xt::xtensor<double, 2> coordinate_dofs,
-                    const std::int32_t index, const std::int32_t perm,
-                    std::shared_ptr<const dolfinx::fem::FiniteElement> element,
-                    const dolfinx::fem::CoordinateElement& cmap)
+void pull_back(xt::xtensor<double, 3>& J, xt::xtensor<double, 3>& K,
```
Comment (jorgensd) [inline cpp/utils.hpp:604]: Some documentation of what this function does, what it computes, and what it requires as input and shape of inputs would be nice:D

File: cpp/utils.hpp [hunk 7]
```diff
@@ -690,6 +675,49 @@ get_basis_functions(xt::xtensor<double, 3>& J, xt::xtensor<double, 3>& K,
       detJ[p] = cmap.compute_jacobian_determinant(_J);
     }
   }
+}
+//-----------------------------------------------------------------------------
+xt::xtensor<double, 3>
```
Comment (jorgensd) [inline cpp/utils.hpp:680]: Add docs

File: cpp/utils.hpp [hunk 8]
```diff
@@ -742,6 +770,66 @@ get_basis_functions(xt::xtensor<double, 3>& J, xt::xtensor<double, 3>& K,
   }
   return basis_array;
 }
+//-----------------------------------------------------------------------------
```
Comment (jorgensd) [inline cpp/utils.hpp:773]: Add docs

File: python/dolfinx_contact/one_sided/nitsche_rigid_surface.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    midpoint_tree = _geometry.create_midpoint_tree(mesh, fdim, contact_facets)
+
+    # This function returns Pi(x) - x, where Pi(x) is the closest point projection
+    def gap(x):
+        dist_vec_array = np.zeros((gdim, x.shape[1]))
+        facets = _geometry.compute_closest_entity(master_bbox, midpoint_tree, mesh, np.transpose(x))
+        for i in range(x.shape[1]):
+            xi = x[:, i]
+            facet = facets[i]
+            if (np.argwhere(np.array(contact_facets) == facet)).shape[0] > 0:
```
Comment (jorgensd) [inline python/dolfinx_contact/one_sided/nitsche_rigid_surface.py:179]: Some documentation of the logic here would be good

File: python/dolfinx_contact/plotting.py [hunk 10]
```diff
@@ -1,4 +1,4 @@
-# Copyright (C) 2021 Jørgen S. Dokken and Sarah Roggendorf
```
Comment (jorgensd) [inline python/dolfinx_contact/plotting.py:1]: 😄

File: python/dolfinx_contact/unbiased/nitsche_unbiased.py [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+    def A(x, A):
+        u.vector[:] = x.array
+        u_opp_0 = contact.pack_u_contact(0, u._cpp_object, gap_0)
+        u_opp_1 = contact.pack_u_contact(1, u._cpp_object, gap_1)
+        u_0 = dolfinx_cuas.pack_coefficients([u], entities_0)
+        u_1 = dolfinx_cuas.pack_coefficients([u], entities_1)
+        c_0 = np.hstack([coeff_0, u_0, u_opp_0])
+        c_1 = np.hstack([coeff_1, u_1, u_opp_1])
+        contact.assemble_matrix(A, [], 0, kernel_jac, c_0, consts)
+        # To create onesided contact assembly remove comment
```
Comment (jorgensd) [inline python/dolfinx_contact/unbiased/nitsche_unbiased.py:194]: I guess these are strictly not true (as we also need to modify the kernel). So I guess we can just remove the comments here and below

**Review by jorgensd — CHANGES_REQUESTED** [review]

File: cpp/Contact.hpp (hunk 2 above)
Comment (SarahRo) [reply cpp/Contact.hpp:1385]: This will be addressed separately. See #26.

**Pull request merged**